### PR TITLE
allow as_set to represent relationals more accurately to not exclude +/-oo

### PR DESCRIPTION
**Pull request by smichr**
#### Release Notes

NO ENTRY

### CODE REVIEW

File: sympy/integrals/integrals.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 from sympy.core.logic import fuzzy_bool
 from sympy.core.mul import Mul
 from sympy.core.numbers import oo, pi
-from sympy.core.relational import Ne
+from sympy.core.relational import Ne, Relational
```
Comment (smichr) [inline sympy/integrals/integrals.py:11]:
```suggestion
from sympy.core.relational import Ne
```

File: sympy/logic/boolalg.py [hunk 2]
```diff
@@ -144,36 +143,36 @@ def as_set(self):
...
         >>> Or(x < -2, 2 < x).as_set()
-        Union(Interval.open(-oo, -2), Interval.open(2, oo))
+        Union({-oo, oo}, Interval.open(-oo, -2), Interval.open(2, oo))
 
         """
         from sympy.calculus.util import periodicity
+        from sympy.core.numbers import oo
         from sympy.core.relational import Relational
+        from sympy.sets.sets import Union
+        from sympy.core.symbol import Symbol, Dummy
```
Comment (smichr) [inline sympy/logic/boolalg.py:157]:
```suggestion
```

File: sympy/logic/boolalg.py [hunk 3]
```diff
@@ -144,36 +143,36 @@ def as_set(self):
...
+                    s = r._eval_as_set()
+                    if s in (S.EmptySet, S.UniversalSet, S.Reals
+                            ) or s == Union({-oo, oo}, S.Reals):
+                        reps[r] = s.as_relational(x)
+                        continue
+                    raise NotImplementedError(filldedent('''
+                        as_set is not implemented for relationals
+                        with periodic solutions
+                        '''))
+            return self.subs(reps)._eval_as_set()
```
Comment (smichr) [inline sympy/logic/boolalg.py:173]:
```suggestion
            if x.kind is NumberKind:
                reps = {}
                for r in self.atoms(Relational):
                    if periodicity(r, x) not in (0, None):
                        s = r._eval_as_set()
                        if s in (S.EmptySet, S.UniversalSet, S.Reals):
                            reps[r] = s.as_relational(x)
                            continue
                        raise NotImplementedError(filldedent('''
                            as_set is not implemented for relationals
                            with periodic solutions
                            '''))
                return self.subs(reps)._eval_as_set()
            return self._eval_as_set()
```

File: sympy/logic/boolalg.py [hunk 4]
```diff
@@ -161,7 +162,8 @@ def as_set(self):
                 for r in self.atoms(Relational):
                     if periodicity(r, x) not in (0, None):
                         s = r._eval_as_set()
-                        if s in (S.EmptySet, S.UniversalSet, S.Reals):
+                        if s in (S.EmptySet, S.UniversalSet, S.Reals
+                                ) or s == Union({-oo, oo}, S.Reals)):
```
Comment (smichr) [inline sympy/logic/boolalg.py:166]:
```suggestion
                                ) or s == Union({-oo, oo}, S.Reals):
```

File: sympy/logic/boolalg.py [hunk 5]
```diff
@@ -10,7 +10,6 @@
 from sympy.core.compatibility import ordered, as_int
 from sympy.core.decorators import sympify_method_args, sympify_return
 from sympy.core.function import Application, Derivative
-from sympy.core.kind import NumberKind
 from sympy.core.numbers import Number
```
Comment (smichr) [inline sympy/logic/boolalg.py:13]:
```suggestion
from sympy.core.kind import NumberKind
from sympy.core.numbers import Number
```

**Comment by smichr** [comment]
Although Interval is supposed to be a subset or Reals, would it make more sense to introduce `Extended` or `ExtendedInterval` that could contain the infinite bound? `Extended` could replace `Interval` and then the new `Interval` would subclass from `Extended` and have the constraint that infinities would not be contained. I would want `isinstance(foo, Interval)` to return True for either class.

File: sympy/functions/elementary/piecewise.py [hunk 6]
```diff
@@ -968,11 +951,18 @@ def _eval_rewrite_as_ITE(self, *args, **kwargs):
...
                             ''' % c))
-                if byfree[x] in (S.UniversalSet, S.Reals):
-                    # collapse the ith condition to True and break
-                    args[i] = list(args[i])
-                    c = args[i][1] = True
-                    break
+                else:
+                    # remove extraneous infinities
+                    cs -= {S.Infinity, S.NegativeInfinity}
+                    # update x
```
Comment (smichr) [inline sympy/functions/elementary/piecewise.py:957]: how do we know that len(free) was 1 and x is unique?

File: doc/src/tutorial/calculus.rst [hunk 7]
```diff
@@ -211,7 +211,7 @@ definite integrals.  Here is a sampling of some of the power of ``integrate``.
     ⌡
     0
     >>> integ.doit()
-    ⎧ Γ(y + 1)    for re(y) > -1
+    ⎧ Γ(y + 1)    for y > -1
```
Comment (smichr) [inline doc/src/tutorial/calculus.rst:214]: the integration variable changed, too, from `x` to `y`!

File: sympy/integrals/integrals.py [hunk 8]
```diff
@@ -1523,7 +1539,7 @@ def integrate(*args, meijerg=None, conds='piecewise', risch=None, heurisch=None,
     in interactive sessions and should be avoided in library code.
 
     >>> integrate(x**a*exp(-x), (x, 0, oo)) # same as conds='piecewise'
-    Piecewise((gamma(a + 1), re(a) > -1),
+    Piecewise((gamma(a + 1), a > -1),
         (Integral(x**a*exp(-x), (x, 0, oo)), True))
```
Comment (smichr) [inline sympy/integrals/integrals.py:1543]: the second integral changed from x to a

File: sympy/integrals/integrals.py [hunk 9]
```diff
@@ -559,10 +561,24 @@ def try_meijerg(function, xab):
...
+                                # not having them in Piecewise doesn't allow
+                                # x < oo to evaluate, so we handle it here
+                                # before creating the Piecewise
+                                free = cond.free_symbols
+                                d = None
+                                if len(free) == 1:
+                                    x = free.pop()
+                                    if not x.is_real:
+                                        d = Dummy(real=True)
+                                        cond = cond.xreplace({x: d})
```
Comment (smichr) [inline sympy/integrals/integrals.py:575]:
```suggestion
                                if len(free) == 1:
                                    cx = free.pop()
                                    if not cx.is_real:
                                        d = Dummy(real=True)
                                        cond = cond.xreplace({cx: d})
```

File: sympy/integrals/integrals.py [hunk 10]
```diff
@@ -559,10 +561,24 @@ def try_meijerg(function, xab):
...
                                 ret = Piecewise(
                                     (f, cond),
                                     (self.func(
                                     function, (x, a, b)), True))
+                                if d:
+                                    ret = ret.xreplace({d: x})
```
Comment (smichr) [inline sympy/integrals/integrals.py:581]:
```suggestion
                                    ret = ret.xreplace({d: cx})
```

**Comment by oscarbenjamin** [comment]
> Although Interval is supposed to be a subset or Reals, would it make more sense to introduce `Extended` or `ExtendedInterval` that could contain the infinite bound? `Extended` could replace `Interval` and then the new `Interval` would subclass from `Extended` and have the constraint that infinities would not be contained. I would want `isinstance(foo, Interval)` to return True for either class.

That sounds reasonable although maybe `Interval` should just be allowed to include infinities. The code seems to be somewhat complicated by the seemingly arbitrary requirement not to have the infinities included. It makes sense to exclude infinities from Reals but I don't see why Interval should not just be able to include infinity.

**Comment by smichr** [comment]
> I don't see why Interval should not just be able to include infinity.

@asmeurer had said that this would be a backwards incompatible change since Interval is supposed to be a subset of Reals.

**Pull request closed without merging**